File: weyl-core/src/main/java/io/quartic/weyl/core/model/StaticSchema.java
package io.quartic.weyl.core.model;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.quartic.common.SweetStyle;
import org.immutables.value.Value;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

@SweetStyle
@Value.Immutable
@JsonSerialize(as = StaticSchemaImpl.class)
@JsonDeserialize(as = StaticSchemaImpl.class)
public interface StaticSchema {
    Optional<AttributeName> titleAttribute();
    Optional<AttributeName> primaryAttribute();
    Optional<AttributeName> imageAttribute();
    List<AttributeName> blessedAttributes();    // Order is important here
    Set<AttributeName> categoricalAttributes();

    Map<AttributeName, AttributeType> attributeTypes();
}

File: weyl-core/src/main/kotlin/io/quartic/weyl/core/model/StaticSchema.kt
package io.quartic.weyl.core.model

// TODO: remove default values once everything in Kotlin
data class StaticSchema @JvmOverloads constructor(
    val titleAttribute: AttributeName? = null,
    val primaryAttribute: AttributeName? = null,
    val imageAttribute: AttributeName? = null,
    val blessedAttributes: List<AttributeName> = emptyList(),       // Order is important here
    val categoricalAttributes: Set<AttributeName> = emptySet(),
    val attributeTypes: Map<AttributeName, AttributeType> = emptyMap()
)
